File: reorder-list.py
from __future__ import annotations
from typing import Any, Optional


class ListNode:
    def __init__(self, val: Any = 0, next: Optional[ListNode] = None):
        self.val = val
        self.next = next


class Solution:
    def reorderList(self, head: Optional[ListNode]) -> None:
        global ptr

        if head is None or head.next is None:
            return None

        fast = slow = head
        while fast:
            try:
                fast = fast.next.next
                slow = slow.next
            except:
                break

        mid = slow
        ptr = head

        def dfs(node: Optional[ListNode]):
            global ptr

            if node is None:
                return

            dfs(node.next)

            ptr_next = ptr.next
            ptr.next = node
            node.next = ptr_next
            ptr = ptr_next

        dfs(mid.next)
        mid.next = None

File: reorder-list.cpp
using namespace std;


struct ListNode {
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};

class Solution {
public:
    void reorderList(ListNode* head) {
		if ((head == nullptr) || (head->next == nullptr)) return;

		ListNode *slow = head;
		ListNode *fast = head;

		while (fast != nullptr) {
			if ((fast->next != nullptr) && (fast->next->next != nullptr)) {
				fast = fast->next->next;
				slow = slow->next;
			} else {
				break;
			}
		}

		ListNode *mid = slow;
		ListNode *ptr = head;

		mid = dfs(mid->next, ptr);
		mid->next = nullptr;
	}

	ListNode *dfs(ListNode *node, ListNode *ptr) {
		if (node == nullptr) return ptr;

		ptr = dfs(node->next, ptr);

		ListNode *ptr_next = ptr->next;
		ptr->next = node;
		node->next = ptr_next;
		return ptr_next;
    }
};
